### PR TITLE
Support BETWEEN in Delta check constraints

### PR DESCRIPTION
**Pull request by pajaks**
## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: plugin/trino-delta-lake/src/main/antlr4/io/trino/plugin/deltalake/expression/SparkExpressionBase.g4 [hunk 1]
```diff
@@ -30,6 +30,7 @@ booleanExpression
     : valueExpression predicate[$valueExpression.ctx]?   #predicated
     | left=booleanExpression AND right=booleanExpression #and
     | left=booleanExpression OR right=booleanExpression #or
+    | operand=valueExpression operator=(BETWEEN | NOT_BETWEEN) leftBound=valueExpression AND rightBound=valueExpression #between
```
Comment (ebyhr) [inline plugin/trino-delta-lake/src/main/antlr4/io/trino/plugin/deltalake/expression/SparkExpressionBase.g4:33]: Can we use the same style of SqlBase.g4? (move to under `#comparison` and use:
```sql
| NOT? BETWEEN lower=valueExpression AND upper=valueExpression        #between
```

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeCheckConstraintCompatibility.java [hunk 2]
```diff
@@ -89,6 +89,12 @@ public static Object[][] checkConstraints()
                 {"a INT, b INT", "a = b % 2", "1, 5", row(1, 5), "1, 6"},
                 {"a INT, b INT", "a = b & 5", "1, 3", row(1, 3), "1, 4"},
                 {"a INT, b INT", "a = b ^ 5", "6, 3", row(6, 3), "6, 4"},
+                // Between
+                {"a INT", "a BETWEEN 1 AND 10", "1", row(1), "0"},
+                {"a INT", "a BETWEEN 1 AND 10", "10", row(10), "11"},
+                {"a INT", "a NOT BETWEEN 1 AND 10", "0", row(0), "1"},
+                {"a INT", "a NOT BETWEEN 1 AND 10", "11", row(11), "10"},
+                {"a INT, b INT, c INT", "a BETWEEN b AND c", "5, 1, 10", row(5, 1, 10), "11, 1, 10"},
```
Comment (ebyhr) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeCheckConstraintCompatibility.java:97]: Let's add another case that the lower expression is NULL. e.g.  `a BETWEEN NULL AND 10`

**Review by ebyhr — APPROVED** [review]

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/expression/NullLiteral.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    public <R, C> R accept(SparkExpressionTreeVisitor<R, C> visitor, C context)
+    {
+        return visitor.visitNullLiteral(this, context);
+    }
+
+    @Override
+    public String toString()
+    {
+        return toStringHelper(this)
+                .toString();
```
[the remark this answers is not in the document]
Comment (pajaks) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/expression/NullLiteral.java:31]: All other Literal classes produce `ClassName{value=x}` so I think `NullLiteral` should follow this convention. WDYT?

**Comment by pajaks** [comment]
@ebyhr Can you run with secrets?

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeCheckConstraintCompatibility.java [hunk 4]
```diff
@@ -98,6 +104,32 @@ public static Object[][] checkConstraints()
...
+        onDelta().executeQuery("ALTER TABLE default." + tableName + " ADD CONSTRAINT a_constraint CHECK (a > 1)");
+
+        try {
+            // Values which produces unknown conditions are treated as FALSE by DELTA specification and as TRUE by Trino according to SQL standard
+            // https://github.com/delta-io/delta/issues/1714
+            assertThatThrownBy(() -> onDelta().executeQuery("INSERT INTO default." + tableName + " VALUES (null)"))
+                    .hasMessageMatching("(?s).* CHECK constraint .* violated by row with values.*");
+            onTrino().executeQuery("INSERT INTO delta.default." + tableName + " VALUES (null)");
+            assertThat(onTrino().executeQuery("SELECT * FROM delta.default." + tableName))
+                    .containsOnly(row(new Object[] {null}));
```
Comment (ebyhr) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeCheckConstraintCompatibility.java:126]: nit: There's no need to declare as an array. Just `(Object) null` is sufficient.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/expression/BetweenPredicate.java [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        BetweenPredicate that = (BetweenPredicate) o;
+        return operator == that.operator &&
+                value == that.value &&
```
Comment (krvikash) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/expression/BetweenPredicate.java:92]: Do we intend to check the memory reference here? or content of value?

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeCheckConstraintCompatibility.java (hunk 2 above)
Comment (krvikash) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeCheckConstraintCompatibility.java:97]: can we have this test for `a NOT BETWEEN b AND c`?

Comment (ebyhr) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeCheckConstraintCompatibility.java:97]: The existing cases are enough in my opinion.

**Pull request merged**